### PR TITLE
[FREELDR] ntldr: Create and use CmpFreeEx()

### PR DESCRIPTION
**Pull request by SergeGautherie**
## Purpose

More secure code.

## Proposed changes

- Create and use CmpFreeEx()

### CODE REVIEW

File: boot/freeldr/freeldr/ntldr/registry.c [hunk 1]
```diff
@@ -52,6 +52,16 @@ CmpAllocate(
...
+    _In_ ULONG Quota,
+    _In_ ULONG Tag)
+{
+    UNREFERENCED_PARAMETER(Quota);
+    FrLdrHeapFree(Ptr, Tag);
+}
+
 VOID
 NTAPI
 CmpFree(
```
Comment (SergeGautherie) [inline boot/freeldr/freeldr/ntldr/registry.c:67]: I assume `HvInitialize()` cannot be modified, can it?
https://github.com/reactos/reactos/blob/c6c6c62cb5b44468aad38b0604339a97f80b1b69/sdk/lib/cmlib/cmlib.h#L369

Comment (HBelusca) [reply boot/freeldr/freeldr/ntldr/registry.c:67]: It cannot.

**Review by HBelusca — CHANGES_REQUESTED** [review]
This CmpFreeEx thing doesn't help since, as you've noticed, cmlib's free callback doesn't take a tag, and modifying this would break NT compatibility. Only thing is that it adds extra code to freeldr for no reason, which is already under tight size constraints.
The sizeof modifications aren't necessary as well.
So please revert all that. The only stuff interesting (that I'll review) are these "impossible cases" you say are present.

**Comment by SergeGautherie** [comment]
> Only thing is that it adds extra code to freeldr for no reason

The reason is to check the tags... (consistency and safety)

> The sizeof modifications aren't necessary as well.

Sure, they are only for 'don't use types if not needed'.

> revert all that

You sure?

**Comment by SergeGautherie** [comment]
> Only thing is that it adds extra code to freeldr for no reason

The reason is to check the tags... (consistency and safety)

> The sizeof modifications aren't necessary as well.

Sure, they are only for 'don't use types if not needed' and consistency within this file.

> revert all that

2nd attempt, disabled by default: at least, self-documentation and could be enabled when touching this code for example.
Do you confirm that?

> "impossible cases"

Split as a separate commit, to ease review.

File: boot/freeldr/freeldr/ntldr/wlregistry.c [hunk 2]
```diff
@@ -753,33 +773,36 @@ WinLdrAddDriverToList(
...
         }
+#endif
         if (DriverNode->Name.Buffer)
-        {
-            CmpFree(DriverNode->Name.Buffer,
-                    DriverNode->Name.MaximumLength);
-        }
-        CmpFree(DriverNode, sizeof(BOOT_DRIVER_NODE));
+            CmpFreeEx(DriverNode->Name.Buffer, DriverNode->Name.MaximumLength, TAG_CM);
+        CmpFreeEx(DriverNode, sizeof(*DriverNode), TAG_CM);
```
Comment (HBelusca) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:805]: Please revert all these.

File: boot/freeldr/freeldr/ntldr/wlregistry.c [hunk 3]
```diff
@@ -753,33 +773,36 @@ WinLdrAddDriverToList(
...
         }
         if (DriverEntry->FilePath.Buffer)
         {
-            CmpFree(DriverEntry->FilePath.Buffer,
-                    DriverEntry->FilePath.MaximumLength);
+            CmpFreeEx(DriverEntry->FilePath.Buffer,
+                      DriverEntry->FilePath.MaximumLength,
+                      TAG_WLDR_NAME);
         }
+#endif
```
Comment (HBelusca) [inline boot/freeldr/freeldr/ntldr/wlregistry.c:802]: Confirmed impossible cases, you can delete them.

**Comment by HBelusca** [comment]
Please don't bring such similar "improvements" (that are not really ones) to the cmlib and ntoskrnl/cm(boot.c) files.

**Pull request closed without merging**